src/number_theory/cyclotomic/rat.lean

```diff
@@ -0,0 +1,154 @@
+enough and less cumbersome to use than the previous lemmas. -/
+lemma discr_prime_pow_eq_unit_mul_pow' [is_cyclotomic_extension {p ^ k} ℚ K]
+  (hζ : is_primitive_root ζ ↑(p ^ k)) :
+  ∃ (u : ℤˣ) (n : ℕ), discr ℚ (hζ.sub_one_power_basis ℚ).basis = u * p ^ n :=
+begin
+  rw [hζ.discr_zeta_eq_discr_zeta_sub_one.symm],
+  exact discr_prime_pow_eq_unit_mul_pow hζ (cyclotomic.irreducible_rat (p ^ k).pos)
+    (cyclotomic.irreducible_rat p.pos)
+end
+
+/-- If `K` is a `p ^ k`-th cyclotomic extension of `ℚ`, then `(adjoin ℤ {ζ})` is the
+integral closure of `ℤ` in `K`. -/
+lemma is_integral_closure_adjoing_singleton_of_prime_pow
+  [hcycl : is_cyclotomic_extension {p ^ k} ℚ K] (hζ : is_primitive_root ζ ↑(p ^ k)) :
+  is_integral_closure (adjoin ℤ ({ζ} : set K)) ℤ K :=
+begin
+  refine ⟨subtype.val_injective, λ x, ⟨λ h, ⟨⟨x, _⟩, rfl⟩, _⟩⟩,
+  { let B := hζ.sub_one_power_basis ℚ,
+    have hint : is_integral ℤ B.gen :=  is_integral_sub (hζ.is_integral (p ^ k).pos)
+      is_integral_one,
+    have H := discr_mul_is_integral_mem_adjoin ℚ hint h,
+    obtain ⟨u, n, hun⟩ := discr_prime_pow_eq_unit_mul_pow' hζ,
+    rw [hun] at H,
+    replace H := subalgebra.smul_mem _ H u.inv,
+    rw [← smul_assoc, ← smul_mul_assoc, units.inv_eq_coe_inv, coe_coe, zsmul_eq_mul,
+      ← int.cast_mul, units.inv_mul, int.cast_one, one_mul,
+      show (p : ℚ) ^ n • x = ((p : ℕ) : ℤ) ^ n • x, by simp [smul_def]] at H,
+    unfreezingI { cases k },
+    { haveI : is_cyclotomic_extension {1} ℚ K := by simpa using hcycl,
+      have : x ∈ (⊥ : subalgebra ℚ K),
+      { rw [singleton_one ℚ K],
+        exact mem_top },
+      obtain ⟨y, hy⟩ := mem_bot.1 this,
+      rw [← hy] at h,
+      replace h := (is_integral_algebra_map_iff (algebra_map ℚ K).injective).1 h,
+      obtain ⟨z, hz⟩ := is_integrally_closed.is_integral_iff.1 h,
+      rw [← hy, ← hz, ← is_scalar_tower.algebra_map_apply],
+      exact subalgebra.algebra_map_mem  _ _ },
+    { have hmin : (minpoly ℤ B.gen).is_eisenstein_at (submodule.span ℤ {((p : ℕ) : ℤ)}),
+      { have h₁ := minpoly.gcd_domain_eq_field_fractions ℚ hint,
+        have h₂ := hζ.minpoly_sub_one_eq_cyclotomic_comp
+          (cyclotomic.irreducible_rat (p ^ _).pos),
+        rw [is_primitive_root.sub_one_power_basis_gen] at h₁,
+        rw [h₁, ← map_cyclotomic_int, show int.cast_ring_hom ℚ = algebra_map ℤ ℚ, by refl,
+          show ((X + 1)) = map (algebra_map ℤ ℚ) (X + 1), by simp, ← map_comp] at h₂,
+        rw [is_primitive_root.sub_one_power_basis_gen, map_injective (algebra_map ℤ ℚ)
+          ((algebra_map ℤ ℚ).injective_int) h₂],
+        exact cyclotomic_prime_pow_comp_X_add_one_is_eisenstein_at _ _ },
+      refine adjoin_le _ (mem_adjoin_of_smul_prime_pow_smul_of_minpoly_is_eiseinstein_at
+        (nat.prime_iff_prime_int.1 hp.out) hint h H hmin),
+      simp only [set.singleton_subset_iff, set_like.mem_coe],
+      exact subalgebra.sub_mem _ (self_mem_adjoin_singleton ℤ _) (subalgebra.one_mem _) } },
+  { rintro ⟨y, rfl⟩,
```

Comment:
maybe it's worth `swap/focus_on_goal` early so you don't have to have all of the rest of the proof indented